File: ocaml/11-20/20.ml
(* Remove the K'th element from a list. (easy)

The first element of the list is numbered 0, the second 1,...
*)

let rec remove_at n = function
  | [] -> []
  | x::xs -> if n = 0 then xs else x::(remove_at (n-1) xs)


let _ = assert ((remove_at 1 ["a";"b";"c";"d"]) = ["a"; "c"; "d"])

File: prolog/11-20/20.pl
/*
Remove the K'th element from a list.
    Example:
    ?- remove_at(X,[a,b,c,d],2,R).
    X = b
    R = [a,c,d]
*/

remove_at(X, [X|T], 1, Seen, Zs) :- 
  reverse(Seen, Rs), 
  append(Rs, T, Zs). 
remove_at(X, [H|T], N, Seen, Zs) :-
  N1 is N - 1,
  remove_at(X, T, N1, [H|Seen], Zs).

remove_at(_, [], _, []).
remove_at(X, [X|T], 1, T).
remove_at(X, [H|T], N, Zs) :-
  N > 1,  
  N1 is N - 1,
  remove_at(X, T, N1, [H], Zs).





